File: 200_number_of_islands/DFS.py
class Solution:
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        """
        Complexity: O(row*col), search all element once
        Space: O(row*col)
        Find connected component.
        """
        def DFS(r, c, visited, row, col):
            rownbr = [-1,0,1,0]
            colnbr = [0,-1,0,1]
            visited[r][c] = True
            for k in range(4): # try not to use loop if the number is small
                new_r = r + rownbr[k]
                new_c = c + colnbr[k]
                if new_r >= 0 and new_c >= 0 and new_r < row and new_c < col and not visited[new_r][new_c] and grid[new_r][new_c] == '1':
                    DFS(new_r, new_c, visited, row, col)
            
        row = len(grid)
        if row == 0:
            return 0
        col = len(grid[0])
        if col == 0:
            return 0
        
        visited = [[False for i in range(col)] for j in range(row)]
        count = 0
        for i in range(row):
            for j in range(col):
                if not visited[i][j] and grid[i][j] == '1':
                    DFS(i, j, visited, row, col)
                    count += 1
        return count

File: 841_Keys_and_Rooms/dfs.cpp
class Solution {
public:
    /*
     * Complexity: O(V+E)
     * Space: O(V)
     */
    void dfs(bool& ans, int room, const vector<vector<int>>& rooms, vector<bool>& visited) {
        visited[room] = true;
        // no false
        if(find(visited.begin(), visited.end(), false) == visited.end()){
            ans = true;
            return;
        }
        for(int key : rooms[room]) {
            if(!visited[key]){
                dfs(ans, key, rooms, visited);
            }
            // able to go to every room
            if(ans) return;
        }
    }
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        if(!rooms.size()) return false;
        vector<bool>visited(rooms.size(), false);
        bool ans = false;
        dfs(ans, 0, rooms, visited);
        return ans;
    }
};
